pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+
+	pub(crate) type CurrencySymbolOf<T> = BoundedVec<u8, <T as Config>::MaxStringLength>;
+
+	pub(crate) type CurveParameterTypeOf<T> = <T as Config>::CurveParameterType;
+
+	// TODO: change CurveParameterTypeOf.
+	pub(crate) type PoolDetailsOf<T> =
+		PoolDetails<<T as frame_system::Config>::AccountId, CurveParameterTypeOf<T>, BoundedCurrencyVec<T>>;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The currency used for storage deposits.
+		type DepositCurrency: MutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>;
+		/// The currency used as collateral for minting bonded tokens.
+		type CollateralCurrency: MutateFungibles<Self::AccountId>
+			+ AccountTouch<CollateralAssetIdOf<Self>, Self::AccountId>
+			+ FungiblesMetadata<Self::AccountId>;
+		/// Implementation of creating and managing new fungibles
+		type Fungibles: CreateFungibles<Self::AccountId, AssetId = Self::AssetId>
+			+ DestroyFungibles<Self::AccountId>
+			+ FungiblesMetadata<Self::AccountId>
+			+ FungiblesInspect<Self::AccountId>
+			+ MutateFungibles<Self::AccountId, Balance = CollateralCurrencyBalanceOf<Self>>;
+		/// The maximum number of currencies allowed for a single pool.
+		#[pallet::constant]
+		type MaxCurrencies: Get<u32>;
+		/// The deposit required for each bonded currency.
+
+		#[pallet::constant]
+		type MaxStringLength: Get<u32>;
+
+		/// The deposit required for each bonded currency.
+		#[pallet::constant]
+		type DepositPerCurrency: Get<DepositCurrencyBalanceOf<Self>>;
+
+		/// The base deposit required to create a new pool, primarily to cover the ED of the pool account.
+		#[pallet::constant]
+		type BaseDeposit: Get<DepositCurrencyBalanceOf<Self>>;
+
+		/// The asset id of the collateral currency.
+		type CollateralAssetId: Get<CollateralAssetIdOf<Self>>;
+		/// Who can create new bonded currency pools.
+		type PoolCreateOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+		/// The type used for pool ids
+		type PoolId: Parameter + MaxEncodedLen + From<[u8; 32]> + Into<Self::AccountId>;
+
+		/// The type used for asset ids. This is the type of the bonded currencies.
+		type AssetId: Parameter + Member + FullCodec + TypeInfo + MaxEncodedLen + Saturating + One + Default;
+
+		type RuntimeHoldReason: From<HoldReason>;
+
+		/// The type used for the curve parameters.
+		type CurveParameterType: Parameter
+			+ Member
+			+ FixedPointNumber<Inner = u128>
+			+ TypeInfo
+			+ MaxEncodedLen
+			+ Zero
+			+ CheckedAdd
+			+ CheckedSub;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// Bonded Currency Swapping Pools
+	#[pallet::storage]
+	#[pallet::getter(fn pools)]
```

Comment:
Very minor nitpick, but if you were to call this on the pallet, you would rather call `pool_with_id(id)` rather than `pools(did)`.

Reply:
bcd2ef34cad80d6d3cff3a5bf77024f12cab3a75

Reply:
Yes, you would call it with `pallet_bonded_coins::Pallet::<T>:: pool_with_id(id)`.